CI/CD refactor and switch to github actions

lmmentel opened the pull request:
#### Reference Issues/PRs

- Fixes #1482 
- Fixes #1728 
- Fixes #1803
- Fixes #1832

See also #1150 
Refactoring of github actions workflows started in #1321 

Currently blocked by: 
- [x] #1644
- [x] #1661
- [x] #1662 
- [x] #1792
- [x] #1747
- [x] #1783
- [x] #1803
- [x] update installation instructions in [docs/source/installation.rst](docs/source/installation.rst)


#### What does this implement/fix? Explain your changes.

This is a rather substantial effort involving changes to how `sktime` is build and distributed. The main changes are:

-  Add [pyproject.toml](https://pip.pypa.io/en/stable/reference/build-system/pyproject-toml/) (see [PEP 621](https://www.python.org/dev/peps/pep-0621/) and [PEP 633](https://www.python.org/dev/peps/pep-0633/)]. The intention behind is to collect all package matadata and dependencies in one place.
- Add sets of optional dependencies:
    - `all_extras` all "soft" dependencies
    - `dev` packages useful for development and testing
    - `binder` packages required for binder and running notebook examples
    - `docs` packages required to build the documentation
- Add section to build time dependencies.
- Refactor existing CI/CD pipelines to use the new build configuration.
- Introduced a new CI/CD pipelines for running tests across win, mac, linux and py37, py38, py39 with github actions.
- Removed the obsolete requirements files.
- Include wheels build for macosx on arm64.

**Removed**

- ❌ Drop support for python3.6
- ❌ Drop appveyor and azure pipelines

**High level diagram for the CI/CD workflows**

![sktime-cicd](https://user-images.githubusercontent.com/8989838/149678967-309054eb-e91e-4988-a6cf-321de0108eb8.png)


#### Does your contribution introduce a new dependency? If yes, which one?

It does not introduce any new runtime/install dependencies but it could introduce an optional development depencency/build system such as [poetry](https://python-poetry.org/) to simplify development/build/distribution.

#### What should a reviewer concentrate their feedback on?

This PR introduces changed to the CI/CD pipelines that would hopefully allow to drop ``appveyor`` and ``azure pipelines`` in favor of ``github actions`` which would simplify the current build and maintenance process.

#### Any other comments?


#### PR checklist

##### For all contributions
- [x] I've added myself to the [list of contributors](https://github.com/alan-turing-institute/sktime/blob/main/.all-contributorsrc).
- [x] Optionally, I've updated sktime's [CODEOWNERS](https://github.com/alan-turing-institute/sktime/blob/main/CODEOWNERS) to receive notifications about future changes to these files.
- [x] I've added unit tests and made sure they pass locally.

lmmentel commented:
@mloning if this is a potential way forward it would be great to trigger the CI runs in order to check that everything runs as expected. I tried that locally with ``act`` but some of the steps fail since it's hard to reproduce the build env exactly.

lmmentel commented:
My initial plan is to simplify the pipelines:

- try to run all on ``github actions`` since it support the [main os hosts](https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners)
- use [cibuildwheel](https://github.com/pypa/cibuildwheel) to build wheels across the matrix of win, linux, mac and python 3.6, 3.7, 3.8, 3.9
- publish to PyPI


After testing locally it seems that ``cibuildwheel`` [does not work well with pure python distributions](https://github.com/pypa/cibuildwheel/issues/862), an alternative would be to simply use ``python -m build --sdist --wheel``.

Are there any C extensions for ``sktime``?

fkiraly commented:
minor comment: is `all` (morally) the same as `all_extras` so far? If so, I'd suggest to keep the old name `all_extras` to avoid to break users' CI/CD which may use `pip install sktime[all_extras]` or similar.

fkiraly commented:
> it would be great to trigger the CI runs

Slightly weird, the familiar button to run the CI does not appear. Not sure why, is everything correct?
This is how it looks like to me:
![image](https://user-images.githubusercontent.com/7985502/142787100-b6cc549e-109a-4c22-b675-09b63f2b0966.png)

lmmentel commented:
Hmmm, I'm not sure but maybe the unresolved file conflicts are still blocking? Couldn't find any relevant information in the docs about it so I'm guessing here.

fkiraly commented:
might be - would you mind merge/resolving?

lmmentel commented:
I rebased on the latest `main` and resolved the conflict.

fkiraly commented:
> I rebased on the latest main and resolved the conflict.

Yes, that was it - now I can "approve and run". Good luck with the tests...

fkiraly commented:
uh-oh, you made the linter unhappy, since there are no newlines at the end of the files!

I corrected your terrible mistake 😁 - I hope you don't mind, just so it runs through the interesting part.

lmmentel commented:
> uh-oh, you made the linter unhappy, since there are no newlines at the end of the files!
> I corrected your terrible mistake grin - I hope you don't mind, just so it runs through the interesting part.

Thanks!

lmmentel commented:
I'm having some challenges understanding the build process and in particular the contents on the `setup.py` where a lot is happening. Any chance there exists documentation about it?

fkiraly commented:
>To be clear I understand the code but I'm missing the why.

Same here - I think that only @mloning can answer, he created the original script a couple years ago.
Might be best to ping him on slack if you expect there to be a back/forth of questions.

fkiraly commented:
I think some of it might be copy-pasted from scikit-learn, so it may be worth investigating there, to.

lmmentel commented:
It seems that a lot of the build code is quite incompatible with tools like [build](https://github.com/pypa/build) specifically 

- how cython extensions are (not) declared/discovered
- `setup.py` 
- custom build scripts inside the package [sktime/setup.py](https://github.com/alan-turing-institute/sktime/blob/main/sktime/setup.py)
- more build tools inside the package [sktime/_build_utils](https://github.com/alan-turing-institute/sktime/tree/main/sktime/_build_utils)
- and [build_tools](https://github.com/alan-turing-institute/sktime/tree/main/build_tools) in the project root

It's quite challenging to piece together what's really happening. On top of that both extensions

- ``sktime/distances/elastic_cython.pyx``
- ``sktime/classification/shapelet_based/mrseql/mrseql.pyx``

that seem to require all this extra configuration are marked for removal, and the share the following header:

    # TODO remove in v0.10.0
    # the functionality in this file is depreciated and to be replaced with a version
    # based on numba.

Maybe it is worth to deal with that first since then the build pipeline might be simplified. Correct me if I'm wrong but replacing the current extensions with numba will leave only pure python code without necessity for compilation.

I'll join slack and try to initiate the discussion about this over there. I only joined discord but it seems a bit quiet. Is slack the preferred communication channel?

fkiraly commented:
> I'll join slack and try to initiate the discussion about this over there. I only joined discord but it seems a bit quiet. Is slack the preferred communication channel?

We've been using slack for developer communication, and discord for events mostly, like community meet-up or dev meetings.
User questions and troubleshooting comes in through three channels currently: GitHub discussions, gist, and discord.

We may want to clean up this proliferation of communication tools at some point, but that's just my opinion...

fkiraly commented:
FYI, here's the "remove cython" project: https://github.com/alan-turing-institute/sktime/projects/24
@TonyBagnall runs the board. Once on slack, I'd recommend to ping @mloning, @freddyaboulton (for CI/CD) and @TonyBagnall (for cython removal)

lmmentel commented:
> FYI, here's the "remove cython" project: https://github.com/alan-turing-institute/sktime/projects/24 @TonyBagnall runs the board. Once on slack, I'd recommend to ping @mloning, @freddyaboulton (for CI/CD) and @TonyBagnall (for cython removal)

Great! There is already a thread on slack about this issue (you're tagged as well, provided I used the right handle :wink: )

fkiraly commented:
> There is already a thread on slack about this issue

Yes, I saw this after I posted only.

mloning commented:
Hi @lmmentel, sorry for the slow reply, in case you're not aware of this already, we're having a CI/CD session tomorrow, would be good to discuss your suggested changes there.

I may have missed discussions elsewhere, a few general comments: 
* A lot of the `setup.py` related code is quite old and was originally copied from scikit-learn.
* Moving everything to GitHub Actions would simplify things, but may slow down CI/CD due to the limit on parallel jobs. Moving away from appveyor would be good.  
* Using `cibuildwheel` seems like a good idea.

lmmentel commented:
@mloning thanks for the context! I wasn't aware of the CI/CD session, but It would be great if I could join the discussion. Although I'm trying to go step by step, the changes I'm exploring here are rather severe so it would definitely be useful to discuss and get broader perspective. 

I don't have any numbers yet in terms of timing of different CI workflows but that would be interesting to collect.

fkiraly commented:
> but It would be great if I could join the discussion

on discord, tomorrow as part of the community collaboration session starting at 1pm UTC. We'll might have some introductions to new community members and a stand-up first, so I'd expect the CI/CD overview to start around 1:15pm.

mloning commented:
There's one other reason why relying on multiple platforms may be preferable over using a single one: sometimes CI platforms break down or make changes which break our pipelines and it may be easier to have some other platform to fall back on in these cases (makes it easier to isolate issues, prevents total standstill of dev work, etc).

lmmentel commented:
@mloning that is true although I would not underestimate the maintenance effort required to support multiple different technologies for essentially doing the same job. I like to think about it in terms of tradeoffs rather that optimal solutions since with a particular technology choice you gain some desired feature while incurring some debt.

mloning commented:
@lmmentel yes good point!

mloning commented on .github/workflows/build-wheels-upload-to-pypi.yml:
```diff
@@ -0,0 +1,37 @@
+        run: python -m pip install cibuildwheel
+
+      - name: Build wheels
+        run: python -m cibuildwheel --output-dir wheelhouse
```
I think we would also like to run all tests on the built wheel before uploading it to PyPI.

lmmentel replied:
Great, `cibuildhweel` exposes configuration options to run tests on the wheels, I tested a few options but couldn't make it work yet. Probably need a bit more massaging.

mloning commented on .github/workflows/build-wheels-upload-to-pypi.yml:
```diff
@@ -0,0 +1,37 @@
+        run: python -m cibuildwheel --output-dir wheelhouse
+
+      - name: Publish package to PyPI
+        if: github.event_name == 'push' && startsWith(github.ref, 'refs/tags')
```
Isn't this action only triggered when a release is published on GitHub (see lines 3-5)?

lmmentel replied:
Yup, this is probably redundant. This is the part is still a draft since for obvious reasons it was not run/tested yet. I know pypi has a test server where we could try it. I'm also considering turning this into a separate PR not to pack too many things into this one, any thoughts?

mloning commented on Makefile:
```diff
@@ -23,7 +23,7 @@ install: ## Install for the current user using the default python command
 
 test: ## Run unit tests
-	pytest --cov-report html --cov=sktime --showlocals --durations=20 --pyargs $(PACKAGE)
+	pytest --cov-report html --cov=sktime --showlocals --durations=20 -n 4 --pyargs $(PACKAGE)
```
Do we need to specify pytest-xdist somewhere to use multiple processes and the `-n` flag here?

mloning commented on .github/workflows/build-and-test.yml:
```diff
@@ -36,63 +36,133 @@ jobs:
+          python -m pip install .[all_extras,dev]
+
+      - name: Show dependecies
+        run: python -m pip list
```
Should we use this instead since we're using conda anyway?
```suggestion
        run: conda list
```

lmmentel replied:
Sure, this is a diagnostic print so whichever is more convenient

mloning commented on .github/workflows/build-and-test.yml:
```diff
@@ -36,63 +36,133 @@ jobs:
+      - name: Show dependecies
+        run: python -m pip list
+
+      - name: Run tests
```
Could we use a reusable GitHub action for the test step and possibly other steps (https://docs.github.com/en/actions/learn-github-actions/reusing-workflows)?

lmmentel replied:
Absolutely, that should be the way to go. At first I'm trying to get things working and the we can focus on refactoring. I've noticed however that it is a bit harder to reason about the workflow and diagnose problems when the configuration is distributed across multiple files.

mloning reviewed (commented):
Hi @lmmentel, thanks again for working on this, I've left a few questions which we can discuss in our meeting next week.

fkiraly commented:
quick Q: with #1792 merged there should no longer be blockers? Or are there still? E.g., #1747?

lmmentel commented:
> quick Q: with #1792 merged there should no longer be blockers? Or are there still? E.g., #1747?

There are a few things to iron out still but it looking good. I cannot consistently reproduce #1747, my guess is that since it's testing predicted values against expectation it might not be a good test. Thanks for your insight in the [comment](https://github.com/alan-turing-institute/sktime/issues/1747#issuecomment-1000819137). I will leave it without changes for now unless I can reproduce the test failure.

I am facing some issues with python 3.9 and windows. Annoying thing is that it is not directly related to sktime but to pystan not supporting windows, see [pystan on windows](https://pystan2.readthedocs.io/en/latest/windows.html). Here are some quotes:

> PyStan requires a working C++ compiler. Configuring such a compiler is typically the most challenging step in getting PyStan running.

> PyStan is tested against the mingw-w64 compiler which works on both Python versions (2.7, 3.x) and supports x86 and x64.

> Due to problems with MSVC template deduction, functions with Eigen library are failing. Until this and other bugs are fixed no support is provided for Windows + MSVC. Currently, no fix is known for this problem, other than to change the compiler to GCC or clang-cl.

I'm trying to get around this with conda but there are still some configuration details I need to tune. I'll make an update once I have the solution pinned down.

lmmentel commented:
@fkiraly I managed to fix the last issue with installing the wheel on windows :tada: , it wasn't a broken wheel but the path mixup in testing. I want to remove the appveyor and azure pipelines integrations if you agree. But all the code other changes should be ready for review.

fkiraly commented:
Ah, I see! A path typo then?

I think we should ask all core devs to review this, since it is so crucial, maybe give it a period of one or two weeks?

Also, currently, "build wheels and push to pypi" fails for macos-10.15 and ubuntu-20.04.

fkiraly commented:
Quick question:
I'm not a GitHub actions expert, so here's a series of potentialyl stupid questions caused by me wanting to be on the safe side.
* how do we avoid accidental release on pypi?
* when is the release workflow triggered, which file defines the trigger? Should not be trigger on "Pr to main", right, because then we would release accidentally any time someone updates a PR? Would you kindly be able to point to file and line where the trigger is defined? As said, I'm not good at GitHub actions but feel this is something myself and others reviewing may want to convince themselves of.
* when we release, is there a step of using test.pypi first, before moving to actual pypi? Can you point to relevant file/line?

fkiraly commented:
Besides - excellent and really crucial work imo that I couldn't have done! Thanks a million times, @lmmentel!

lmmentel commented:
> Ah, I see! A path typo then?

Yup, I tried running the test from sktime root and wrong version was being imported.

> I think we should ask all core devs to review this, since it is so crucial, maybe give it a period of one or two weeks?

I'm happy to walk through the changes if that would be helpful and if it is possible to find a slot that would fit everyone. Maybe one of our regular meetings?

> Also, currently, "build wheels and push to pypi" fails for macos-10.15 and ubuntu-20.04.

Yup, sometimes `macports.org` is not available (usually round midnight) and then downloading `libomp.tbz2` fails. Restarting the job should fix it. 

> how do we avoid accidental release on pypi?

Do you have any specific "accident" scenarios in mind? 

> when is the release workflow triggered, which file defines the trigger? Should not be trigger on "Pr to main", right, because then we would release accidentally any time someone updates a PR? Would you kindly be able to point to file and line where the trigger is defined? As said, I'm not good at GitHub actions but feel this is something myself and others reviewing may want to convince themselves of.

I was testing the release flow with a push trigger since running the flow on the actual release trigger would require creating *a lot* of fake releases. In the cleanup I restored the actual release trigger https://github.com/alan-turing-institute/sktime/blob/2a08ee82150029288ee2488036d0829c82a43598/.github/workflows/build-wheels-upload-to-pypi.yml#L4 The workflow will only be run after someone authorized creates a release manually in github.

> when we release, is there a step of using test.pypi first, before moving to actual pypi? Can you point to relevant file/line?

No there is no deploy to test pypi. Do you think it is needed? If so, what for?

fkiraly commented:
> Do you have any specific "accident" scenarios in mind?

* wrong trigger event being set (push, not release)
* a maintainer following the release process incorrectly (click on wrong buttons in GitHub GUI). Accidental release should be more than one click away, if possible.

> I'm happy to walk through the changes if that would be helpful and if it is possible to find a slot that would fit everyone. Maybe one of our regular meetings?

Yes, how about the Jan 14 one?

lmmentel commented:
> > Do you have any specific "accident" scenarios in mind?
> 
> * wrong trigger event being set (push, not release)
This should be picker up in the review process. I guess CI files won't change that much so little risk here, however special attention should be paid when reviewing changes involving CI.
> * a maintainer following the release process incorrectly (click on wrong buttons in GitHub GUI). Accidental release should be more than one click away, if possible.
> 
I don't have a solution to human error problem except limiting the number of people authorized to push that button :wink: In case of error we can delete the release from pypi, increment version and do it again.


> > I'm happy to walk through the changes if that would be helpful and if it is possible to find a slot that would fit everyone. Maybe one of our regular meetings?
> 
> Yes, how about the Jan 14 one?

Works for me!

fkiraly commented:
>This should be picked up in the review process. I guess CI files won't change that much so little risk here, however special attention should be paid when reviewing changes involving CI.

Yes, makes sense. You should probably add yourself and @freddyaboulton to CODEOWNERS then?

>I don't have a solution to human error problem except limiting the number of people authorized to push that button 😉 In case of error we can delete the release from pypi, increment version and do it again.

Is there a way to have two release buttons, one for `test-pypi` and one for `pypi` proper? With the expectation that the process is (a) release on `test-pypi`, (b) inspect and check, (c) release on `pypi`? Where the technology does not allow a release on `pypi` if a release on `test-pypi` has not happened for the same version and wheels?

lmmentel commented:
> Yes, makes sense. You should probably add yourself and @freddyaboulton to CODEOWNERS then?

OK.

> Is there a way to have two release buttons, one for `test-pypi` and one for `pypi` proper? With the expectation that the process is (a) release on `test-pypi`, (b) inspect and check, (c) release on `pypi`? Where the technology does not allow a release on `pypi` if a release on `test-pypi` has not happened for the same version and wheels?

For extra button or other UI elements you would have to talk to github directly since it's their UI. I'm not sure I understand why would you want do a test release.

fkiraly commented:
>For extra button or other UI elements you would have to talk to github directly since it's their UI. I'm not sure I understand why would you want to do a test release.

The buttons are purely metaphorical, I do not suggest to develop UI elements or hack the GitHub one. Sorry, I should have made myself more clear here.

What I mean, in the current release workflow (see `build_tools/make_release.py`) there are a couple of "test" steps like upload to test `pypi` and install from it. There is a further expectation that there is a round of expert inspection before the actual release. The "button" would be whathever it means to then trigger a GitHub actions workflow with ordinary developer IDE and GitHub user interfaces.

lmmentel commented:
> The buttons are purely metaphorical, I do not suggest to develop UI elements or hack the GitHub one. Sorry, I should have made myself more clear here.

Got it, I didn't catch that from what you wrote.

> What I mean, in the current release workflow (see `build_tools/make_release.py`) there are a couple of "test" steps like upload to test `pypi` and install from it. There is a further expectation that there is a round of expert inspection before the actual release. The "button" would be whathever it means to then trigger a GitHub actions workflow with ordinary developer IDE and GitHub user interfaces.

I checked the file and it seems like a bit of an overkill and insufficient at the same time. Indeed there is a a few steps that build and install: 

```python
        ...
        MakeDist(),
        UploadToTestPyPI(),
        InstallFromTestPyPI(),
        ...
```

but I don't think we should use it as is since it can to more harm than good:

- it will build and upload wheels to pypi **only** for the platform on which the script is invoked which is a bit random, compared to the comprehensive build of wheels in the CI with coverage of all supported platforms and python versions
- it will build and upload to pypi a wheel for a single python version which will use whatever python version is found on the host
- after uploading, downloading and installing the wheel it will only test a single platform/single python version 
 
I don't think I can see any benefit of doing a tests upload of a single (or multiple) wheel to pypi before a release. I'm open to being convinced otherwise but I would suggest to revise the process and consider dropping it.

What I would suggest is to craft a release according to a release process/checklist/script and have all the changes put as  a PR. After that PR is reviewed the responsible person can create the actual release on github which will trigger the workflow that will build the wheels, test **all** of them and upload to pypi.

fkiraly commented:
> I don't think I can see any benefit of doing a tests upload of a single (or multiple) wheel to pypi before a release. I'm open to being convinced otherwise but I would suggest to revise the process and consider dropping it.

Hm, why not? It does capture basic install errors and easy-to-detect breakage across all versions, before it goes on `pypi`. With the `pypi` upload, the clock starts ticking as soon as it's up, i.e., installs all around the world breaking as soon as it's there.
In my imagination - and that's perhaps also what @mloning thought - we should at least rule out issues that do not depend on version and OS.

>What I would suggest is to craft a release according to a release process/checklist/script and have all the changes put as a PR. After that PR is reviewed the responsible person can create the actual release on github which will trigger the workflow that will build the wheels, test all of them and upload to pypi.

How would that be different from the `test.pypi` workflow?

fkiraly commented:
(this is a discussion we should also probably have with more people, in the Jan 14 dev meeting)

lmmentel commented:
> > I don't think I can see any benefit of doing a tests upload of a single (or multiple) wheel to pypi before a release. I'm open to being convinced otherwise but I would suggest to revise the process and consider dropping it.
> 
> Hm, why not? It does capture basic install errors and easy-to-detect breakage across all versions, before it goes on `pypi`. With the `pypi` upload, the clock starts ticking as soon as it's up, i.e., installs all around the world breaking as soon as it's there. In my imagination - and that's perhaps also what @mloning thought - we should at least rule out issues that do not depend on version and OS.

That test is already covered in the CI/CD workflow, after the wheels are build on each platform and python version they are installed and tested. Going for the round trip to pypi and back does not change the wheels so additional checks seem redundant.

 
> > What I would suggest is to craft a release according to a release process/checklist/script and have all the changes put as a PR. After that PR is reviewed the responsible person can create the actual release on github which will trigger the workflow that will build the wheels, test all of them and upload to pypi.
> 
> How would that be different from the `test.pypi` workflow?

I have a feeling that we are not talking about the same things here. I'm referring to the workflow here https://github.com/lmmentel/sktime/blob/enh-add-pyproject-toml-issue-1482/.github/workflows/build-wheels-upload-to-pypi.yml when talking about what happens (automatically) after creating and publishing a release on github. Could you elaborate on what you mean by `test.pypi` workflow.

lmmentel commented:
> (this is a discussion we should also probably have with more people, in the Jan 14 dev meeting)

Would it worth linking to this PR in the invite or announcing it on slack so that people have a chance to read though and get some context before the discussion?

fkiraly commented:
@lmmentel, you have some conflicts with the installation.rst?

lmmentel commented:
> @lmmentel, you have some conflicts with the installation.rst?

Must've been some recent changes, let me rebase.

fkiraly commented on .all-contributorsrc:
```diff
@@ -1419,7 +1419,11 @@
+        "doc",
+        "test",
+        "bug",
+        "maintenance"
```
and `"infra"`, no?

fkiraly commented on docs/source/get_involved/contributing.rst:
```diff
@@ -223,17 +223,11 @@ We use `pytest <https://docs.pytest.org/en/latest/>`__ for unit testing.
 development version of sktime and all extra dependencies.
 
-1. Install all extra requirements from the root directory of sktime:
+1. Install the development version of sktime:
```
perhaps more precisely: "install the development version of sktime, with all developer dependencies"?

Similarly, should the dependency sets be listed anywhere in the install instructions?

lmmentel replied:
There are not all developer dependencies since also docs can be considered dev dependencies, but I've updated the line.

fkiraly commented on docs/source/get_involved/contributing.rst:
```diff
@@ -291,13 +285,13 @@ the ``index.rst`` file.
    .. code:: bash
 
-      pip install -r docs/requirements.txt
+      pip install .[docs]
```
hm, should we cater for the user who has a dev install but not the doc dependencies?
Would this not possibly un- or reinstall the rest?

lmmentel replied:
why not? Contributing to OS thought docs is usually a good start for may people.

fkiraly commented on docs/source/get_involved/contributing.rst:
```diff
@@ -385,20 +374,6 @@ integration services we use.
-|               | thub.com/pypa/ |                                     |
-|               | manylinux>`__) |                                     |
-+---------------+----------------+-------------------------------------+
 | `GitHub       | Build/t        | `.github/workflows/ <https://gi     |
```
the GitHub actions row is no longer accurate - should include the full matrix

fkiraly commented on docs/source/installation.rst:
```diff
@@ -5,12 +5,12 @@ Installation
-* environments with python version 3.6, 3.7, or 3.8.
-* operating systems Mac OS X, Unix-like OS, Windows 8.1 and higher
+* Python versions 3.7, 3.8 and 3.9
+* Operating systems Mac OS X, Unix-like OS, Windows 8.1 and higher
```
should we note python 3.6, and say expressly it's no longer supported?

fkiraly commented on docs/source/installation.rst:
```diff
@@ -64,32 +63,33 @@ Release versions - troubleshooting
+There are several workarounds:
 
-Potential workaround no.1, install manually via ``conda-forge``:
+1. Install manually via ``conda-forge``:
```
changing it to the list may confuse users and suggest they need to execute these sequentially, as opposed to alternatively. I suggest bringing back the original formulation.

fkiraly replied:
this is because "1., 2., 3." is strongly interpreted as a sequence.

fkiraly commented on docs/source/installation.rst:
```diff
@@ -141,15 +144,15 @@ Details for this differ by operating system, and can be found in the `sktime bui
 Typically, the set-up steps needs to be carried out only once per system.
 That is, the steps usually do not need to be followed again on the same system
-when installing an ``sktime`` development version for the second or third time.
+when installing ``sktime`` development version for the second or third time.
```
this is no longer grammatically correct? Suggest revert or alternative.

fkiraly commented on docs/source/installation.rst:
```diff
@@ -369,7 +359,7 @@ You will need Build Tools for Visual Studio 2022
 4. Open the Visual Studio installer
 5. Tick ``Desktop Development with C++``
-6. In the panel to the write make sure ``MSVC v143 VS 2022 C++ x64/x86 build tools (Latest)`` is selected
+6. In the panel make sure ``MSVC v143 VS 2022 C++ x64/x86 build tools (Latest)`` is selected
```
comma and "that" missing: "in the panel, make sure that"

fkiraly commented on pyproject.toml:
```diff
@@ -0,0 +1,102 @@
+version = "0.9.0"
+description = "A unified framework for machine learning with time series"
+license = "BSD-3-Clause"
+authors = [
```
we probably want to add other names here - CC and/or core devs?
We should add at least the other past and present CC members, @TonyBagnall, @mloning, and @aiwalter.
For maintainers, the current CC, @TonyBagnall and @aiwalter.

lmmentel replied:
Would @TonyBagnall @aiwalter and @mloning like their contact emails present as well? If so please provide the information.

lmmentel replied:
Reached out to @aiwalter for email but haven't heard back. I'll resolve this for now to be able to merge and we can make the updates later if requested :)

fkiraly commented on pyproject.toml:
```diff
@@ -0,0 +1,102 @@
+classifiers = [
+    "Intended Audience :: Science/Research",
+    "Intended Audience :: Developers",
+    "License :: OSI Approved",
```
this should be `BSD-3-Clause`, no?

fkiraly commented on pyproject.toml:
```diff
@@ -0,0 +1,102 @@
+    "pytest-cov",
+    "pytest-xdist",
+    "wheel",
+]
```
do we not also require `pydocstyle`, `pycodestyle`, `numpydoc`, and `flake8`? Also, should `numpydoc` not be here too, since it is a `pre-commit` check?

lmmentel replied:
`numpydoc` is under `docs` section. All the rest are configured by pre-commit which installs packages configured in `sktime/.pre-commit-config.yaml` first time you run it.

fkiraly commented on setup.py:
```diff
@@ -5,149 +5,179 @@
+# /astropy_helpers/openmp_helpers.py  # noqa
+
+
 __author__ = ["Markus Löning"]
```
should be GitHub idea and now include @lmmentel

fkiraly reviewed (dismissed):
A monumental achievement!

I've focused inspecting the documentation, binder, `.github/workflows`, `pyproject.toml`, and `setup.py` . All looks sensible, although my experience is not deep here. The changes all look sensible, and the working CI/CD seems to be evidence that this is also functional.

I left some comments, all look minor, but I'd recommend to go throught hem.

Further, one question: what is going on in `benchmarking` - why is that being changed?

lmmentel commented:
> A monumental achievement!
> 
> I've focused inspecting the documentation, binder, `.github/workflows`, `pyproject.toml`, and `setup.py` . All looks sensible, although my experience is not deep here. The changes all look sensible, and the working CI/CD seems to be evidence that this is also functional.
> 
> I left some comments, all look minor, but I'd recommend to go throught hem.
> 
> Further, one question: what is going on in `benchmarking` - why is that being changed?

Thanks for the comments @fkiraly ! I made some changes and followed most of your suggestions. If you spot anything please let me know.

There was a testing issues with `benchmarking` module on macosx as far as I remember. It was related to the testing code hard coding relative paths which resulted in a bit unpredictable behavior. I replaced the hard coded paths with a pytest `tmppath` fixture. I had to understand the code being tested and while reading through the model there were some obvious refactors so I couldn't resist and just made the updates. Should've probably wrote an issue but it was quicker without it :slightly_smiling_face:

TonyBagnall commented:
@fkiraly is there a time line for merging this? I'm happy for it to go in. I guess if it all goes wrong (sure it wont!) we can just reverse it?

fkiraly reviewed (changes requested):
@lmmentel, thanks, looks good!

There is one thing, regarding documentation.
I've merged the big documentation PR, which refactors and moves around much of sktime's documentation, https://github.com/alan-turing-institute/sktime/pull/1707

In this, the duplications in `get_involved/contributing.rst` have been resolved, and content on CI/CD moved to other files. You made changes to `contributing.rst`, where the original content is now in other places due ot #1707:
* changes to section on unit testing, now in `developer_guide/continuous_integration.rst` 
* removing old entries from the CI/CD table, now in `developer_guide/continuous_integration.rst` 
* documentation build (your changes on `pyproject.toml` etc), has moved to `developer_guide/documentation.rst` 

Would you kindly be able to ensure the changes get moved across, and don't accidentally reappear in `get_involved/contributing.rst`?
Sorry for this, I was thinking about which way round it's better - the rationale here is that the *primary* change needs to go in in first, and the *secondary* change needs to adapt to it. #1707 is the primary change, when it comes to docs. Same will apply for my testing framework, the other way round (CI/CD is first).

fkiraly commented:
> @fkiraly is there a time line for merging this? I'm happy for it to go in. I guess if it all goes wrong (sure it wont!) we can just reverse it?

@TonyBagnall, as discussed in the core dev meeting yesterday: asap, as soon as we are happy!

lmmentel commented:
@fkiraly I rebased the branch on top of latest main. I had to do manual edits in several places and unfortunately cannot guarantee that the documentation has no errors since it was not a straighforward change given that same files we modified and then modified content was moved to different files. That makes it hard to track changes. Before merging consider build the docs locally to make sure they are as expected.

lmmentel commented:
@fkiraly should be ready to merge :slightly_smiling_face:

fkiraly reviewed (approved):
Alright!

Changes to the docs look good too, so ready to go.
I'll give it another skim, and then I'll press the button 😄

fkiraly reviewed (changes requested):
actually, some minor things:

* it should be "Martin Walter", not "Marting Walter"
* while you're there, can you write out my first name (Franz) in the author/maintainer fields?
* the `binder` set of dependencies includes `LunarCalendar` - why is that even there? It does not seem to be used for anything.

fkiraly reviewed (dismissed):
Alright, looks all good now!

fkiraly commented on pyproject.toml:
```diff
@@ -0,0 +1,106 @@
+]
+
+binder = [
+    "convertdate>= 2.3.2",
```
```suggestion
```

fkiraly reviewed (approved):
Just removed `convertdate` too - same problem, didn't seem to be used anywhere.
